### PR TITLE
Cross compile kamon-instrumentation-common with scala 3

### PR DESCRIPTION
**Pull request by vaslabs**
Hello,
Thank you for your work on Kamon.

We are recently looking to upgrade to scala 3 and stumbled upon some kamon modules not having scala 3 builds. One in particular is kamon-http4s which depends on kamon-core and kamon-instrumentation-common .

~~It seems easy enough to add scala 3 to the latter. I have already done the work for kamon-http4s , just need an official release if you can and send you a PR on that project too.~~

Please let me know what you think.

Kind regards

EDIT:
just saw this https://github.com/kamon-io/Kamon/issues/1064

I'll have a deeper look

### CODE REVIEW

**Comment by vaslabs** [comment]
I'll put my notes here

The issue seems to be originating from byte buddy:
```java
@Override
                protected ParameterDescription resolve(MethodDescription instrumentedMethod) {
                    ParameterList<?> parameters = instrumentedMethod.getParameters();
                    if (parameters.size() <= index) {
                        throw new IllegalStateException(instrumentedMethod + " does not define an index " + index);
                    } else {
                        return parameters.get(index);
                    }
                }
```

The bytecode (javap -v) output of scala 3 doWork method is
```
  public kamon.context.Context doWork();
    descriptor: ()Lkamon/context/Context;
    flags: ACC_PUBLIC
    Code:
      stack=1, locals=1, args_size=1
         0: getstatic     #22                 // Field kamon/Kamon$.MODULE$:Lkamon/Kamon$;
         3: invokevirtual #25                 // Method kamon/Kamon$.currentContext:()Lkamon/context/Context;
         6: areturn
      LineNumberTable:
        line 77: 0
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0       7     0  this   Lkamon/instrumentation/context/ContextInstrumentationSpec$Target;

```

The scala 2.13 is
```
  public kamon.context.Context doWork();
    descriptor: ()Lkamon/context/Context;
    flags: ACC_PUBLIC
    Code:
      stack=1, locals=1, args_size=1
         0: getstatic     #16                 // Field kamon/Kamon$.MODULE$:Lkamon/Kamon$;
         3: invokevirtual #19                 // Method kamon/Kamon$.currentContext:()Lkamon/context/Context;
         6: areturn
      LineNumberTable:
        line 77: 0
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0       7     0  this   Lkamon/instrumentation/context/ContextInstrumentationSpec$Target;


```

**Comment by vaslabs** [comment]
Found a difference in the bytecode which looks to me like debug symbols (from -g:vars) which is ignored in the scala 3 compilation.
So decompiling with FernFlower we have

### Scala 2.13

```java
public final class InvokeWithCapturedContext {
   @OnMethodExit(
      onThrowable = Throwable.class
   )
   public static void exit(@Enter final Storage.Scope scope) {
      .MODULE$.exit(scope);
   }

   @OnMethodEnter
   public static Storage.Scope enter(@This final HasContext hasContext) {
      return .MODULE$.enter(hasContext);
   }
}
```

## Scala 3

```java
public final class InvokeWithCapturedContext {
   @OnMethodEnter
   public static Storage.Scope enter(HasContext var0) {
      return .MODULE$.enter(var0);
   }

   @OnMethodExit(
      onThrowable = Throwable.class
   )
   public static void exit(Storage.Scope var0) {
      .MODULE$.exit(var0);
   }
}

```

**Comment by vaslabs** [comment]
It also looks like that parameter annotations are kept but:

The accompanying scala 3 `InvokeWithCapturedContext$` that has the singleton implementation:
```java

public final class InvokeWithCapturedContext$ implements Serializable {
   public static final InvokeWithCapturedContext$ MODULE$ = new InvokeWithCapturedContext$();

   private InvokeWithCapturedContext$() {
   }

   private Object writeReplace() {
      return new ModuleSerializationProxy(InvokeWithCapturedContext$.class);
   }

   @OnMethodEnter
   public Storage.Scope enter(@This final HasContext hasContext) {
      return .MODULE$.storeContext(hasContext.context());
   }

   @OnMethodExit(
      onThrowable = Throwable.class
   )
   public void exit(@Enter final Storage.Scope scope) {
      scope.close();
   }
}
```

and the scala 2.13


```java
public final class InvokeWithCapturedContext$ {
   public static final InvokeWithCapturedContext$ MODULE$ = new InvokeWithCapturedContext$();

   @OnMethodEnter
   public Storage.Scope enter(@This final HasContext hasContext) {
      return .MODULE$.storeContext(hasContext.context());
   }

   @OnMethodExit(
      onThrowable = Throwable.class
   )
   public void exit(@Enter final Storage.Scope scope) {
      scope.close();
   }

   private InvokeWithCapturedContext$() {
   }
}

```


The annotations go to the underlying singleton type but not to the top level with the static methods

**Comment by vaslabs** [comment]
Ok I found a fix for this

For scala 3 we need to use `@static` which requires a companion class and gives full java compatibility with static methods.

But, this annotation does not seem to be available for scala 2.13 so we'll need other version specific implementations

**Comment by vaslabs** [comment]
This should be ready to be tested for #1191

File: build.sbt [hunk 1]
```diff
@@ -158,6 +158,11 @@ lazy val `kamon-instrumentation-common` = (project in file("instrumentation/kamo
   .disablePlugins(AssemblyPlugin)
   .enablePlugins(JavaAgent)
   .settings(instrumentationSettings)
+  .settings(
+        crossScalaVersions += `scala_3_version`,
+  ).settings(
+    scalacOptions ++= { if(scalaBinaryVersion.value.startsWith("3.")) Seq("-Xtarget:8") else Seq.empty }
```
Comment (ivantopo) [inline build.sbt:164]: Can you please move this setting to this other file: https://github.com/kamon-io/Kamon/blob/master/project/Build.scala#L167

File: build.sbt [hunk 2]
```diff
@@ -821,6 +826,7 @@ lazy val bundle = (project in file("bundle"))
 
 import com.lightbend.sbt.javaagent.Modules
 import BundleKeys._
+import sbt.Keys.scalacOptions
```
Comment (ivantopo) [inline build.sbt:829]: This will go away when the previous line gets moved

**Review by ivantopo — CHANGES_REQUESTED** [review]
Thank you so much @vaslabs!

There is a tiny change in the `scalacOptions` setting and the syntax of annotations, just to make them easier to read.

**Review by ivantopo — APPROVED** [review]

**Comment by ivantopo** [comment]
Thank you so much for the effort @vaslabs!

**Pull request merged**